azure-storage-cpp-lite/src/storage_url.cpp

```diff
@@ -110,5 +110,92 @@ namespace microsoft_azure {
             return url;
         }
 
+        // The URLs this is going to need to parse are fairly unadvanced.
+        // They'll all be similar to http://blah.com/path1/path2?query1=xxx&query2=xxx
+        // It's assumed they will be pre-encoded.
+        // This is _primarily_ to support the custom MSI endpoint scenario requested by AML.
+        std::shared_ptr<storage_url> parse_url(const std::string& url) {
+            auto output = std::make_shared<storage_url>();
+
+            std::string runningString;
+            std::string qpname; // A secondary buffer for query parameter strings.
+            // 0 = scheme, 1 = hostname, 2 = path, 3 = query
+            // the scheme ends up attached to the hostname due to the way storage_urls work.
+            int segment = 0;
+            for (auto charptr = url.begin(); charptr < url.end(); charptr++) {
+                switch (segment) {
+                    case 0:
+                        runningString += *charptr;
+
+                        // ends up something like "https://"
+                        if (*(charptr - 2) == ':' && *(charptr - 1) == '/' && *charptr == '/')
+                        {
+                            // We've reached the end of the scheme.
+                            segment++;
+                        }
+                        break;
+                    case 1:
+                        if (*charptr == '/')
+                        {
+                        domainfinalchar: // Because these jumps are local, it's arguably still easy to debug.
+                            // Only append the new char if it's the end of the string.
+                            output->set_domain(std::string(runningString));
+                            // empty the buffer, do not append the new char to the string because storage_url handles it for us, rather than checking itself
+                            runningString.clear();
+                            segment++;
+                        }
+                        else
+                        {
+                            runningString += *charptr;
+                            if (charptr == url.end() - 1)
+                                goto domainfinalchar; // Because these jumps are local, it's arguably still easy to debug.
```

Comment:
Actually, this can be wildly simplified, in the same way that segment 0 is handled.

TODO: Simplify this code that way.